content/en/blog/_posts/2022-11-11-retroactive-default-storage-class.md

```diff
@@ -0,0 +1,163 @@
+
+
+### Resource ordering during cluster installation
+
+If a cluster installation tool needed to create resources that required storage, e.g., image registry, it was difficult to get the ordering right. 
+This is because any Pods that required storage would rely on the presence of a default StorageClass and would fail to be created if it wasn't defined.
+
+## What changed
+
+We've changed the PersistentVolume (PV) controller to assign a default StorageClass to any unbound PersistentVolumeClaim that has the storageClassName set to <code>null</code>.
+We've also modified the PersistentVolumeClaim admission within the API server to allow the change of values from unset value to an actual StorageClass name.
+
+### <code>nil</code> vs <code>""</code> for storageClassName - does it matter?
+
+Before this feature was introduced, those values were equal in terms of behavior. Any PersistentVolumeClaim with the storageClassName set to <code>nil</code> or <code>""</code> would bind to an existing PersistentVolume resource with storageClassName also set to <code>nil</code> or <code>""</code>.
+
+With this new feature enabled we wanted to maintain this behavior but also be able to update the StorageClass name.
+With these constraints in mind, the feature changes the semantics of <code>nil</code>. If a default StorageClass is present, <code>nil<code> would translate to "Give me a default" and <code>""</code> would mean "Give me PersistentVolume that also has <code>""</code> StorageClass name." In the absence of a StorageClass, the behavior would remain unchanged.
+
+Summarizing the above, we've changed the semantics of <code>nil<code> so that its behavior depends on the presence or absence of a definition of default StorageClass.
+
+The tables below show all these cases to better describe when PVC binds and when it's StorageClass gets updated.
+
+<table>
+  <thead>
+     <tr>
+        <th colspan="2"></th>
+        <th>PVC storageClassName == <code>""</code></th>
+        <th>PVC storageClassName == <code>nil</code></th>
+     </tr>
+  </thead>
+  <tbody>
+     <tr>
+        <td rowspan="2">Without default class</td>
+        <td>PV storageClassName == <code>""</code></td>
+        <td>binds</td>
+        <td>binds</td>
+     </tr>
+     <tr>
+        <td>PV without storageClassName</td>
+        <td>binds</td>
+        <td>binds</td>
+     </tr>
+     <tr>
+        <td rowspan="2">With default class</td>
+        <td>PV storageClassName == <code>""</code></td>
+        <td>binds</td>
+        <td>class updates</td>
+     </tr>
+     <tr>
+        <td>PV without storageClassName</td>
+        <td>binds</td>
+        <td>class updates</td>
+     </tr>
+  </tbody>
+</table>
+
+## How to use it
+
+If you want to test the feature whilst it's alpha, you need to enable the relevant feature gate in the kube-controller-manager and the kube-apiserver. Use the <code>--feature-gates</code> command line argument:
```

Comment:
```suggestion
If you want to test the feature whilst it's alpha, you need to enable the relevant feature gate
in the kube-controller-manager and the kube-apiserver.
Use the `--feature-gates` command line argument:
```
Main thing here is to avoid HTML where Markdown would work.

Reply:
Fixed.